// mysql2/index.js
'use strict';

const SqlString = require('sqlstring');

const ConnectionConfig = require('mysql2/lib/connection_config.js');
const parserCache = require('mysql2/lib/parsers/parser_cache.js');

const Connection = require('mysql2/lib/connection.js');

exports.createConnection = require('mysql2/lib/create_connection.js');
exports.connect = exports.createConnection;
exports.Connection = Connection;
exports.ConnectionConfig = ConnectionConfig;

const Pool = require('mysql2/lib/pool.js');
const PoolCluster = require('mysql2/lib/pool_cluster.js');
const createPool = require('mysql2/lib/create_pool.js');
const createPoolCluster = require('mysql2/lib/create_pool_cluster.js');

exports.createPool = createPool;

exports.createPoolCluster = createPoolCluster;

exports.createQuery = Connection.createQuery;

exports.Pool = Pool;

exports.PoolCluster = PoolCluster;

exports.createServer = function (handler) {
  const Server = require('mysql2/lib/server.js');
  const s = new Server();
  if (handler) {
    s.on('connection', handler);
  }
  return s;
};

exports.PoolConnection = require('mysql2/lib/pool_connection.js');
exports.authPlugins = require('mysql2/lib/auth_plugins');
exports.escape = SqlString.escape;
exports.escapeId = SqlString.escapeId;
exports.format = SqlString.format;
exports.raw = SqlString.raw;

exports.__defineGetter__(
  'createConnectionPromise',
  () => require('mysql2/promise.js').createConnection,
);

exports.__defineGetter__(
  'createPoolPromise',
  () => require('mysql2/promise.js').createPool,
);

exports.__defineGetter__(
  'createPoolClusterPromise',
  () => require('mysql2/promise.js').createPoolCluster,
);

exports.__defineGetter__('Types', () => require('mysql2/lib/constants/types.js'));

exports.__defineGetter__('Charsets', () =>
  require('mysql2/lib/constants/charsets.js'),
);

exports.__defineGetter__('CharsetToEncoding', () =>
  require('mysql2/lib/constants/charset_encodings.js'),
);

exports.setMaxParserCache = function (max) {
  parserCache.setMaxCache(max);
};

exports.clearParserCache = function () {
  parserCache.clearCache();
};
